src/platform/unix/mod.rs

```diff
@@ -465,19 +464,23 @@ impl OsIpcReceiverSet {
                                 shared_memory_regions));
                     }
                     Err(err) if err.channel_is_closed() => {
-                        hangups.insert(pollfd.fd);
-                        selection_results.push(OsIpcSelectionResult::ChannelClosed(
-                                    *self.fdids.get(&pollfd.fd).unwrap()))
+                        let id = self.fdids.remove(&pollfd.fd).unwrap();
+                        unsafe {
+                            libc::close(pollfd.fd);
+                        }
+                        selection_results.push(OsIpcSelectionResult::ChannelClosed(id))
                     }
                     Err(err) => return Err(err),
                 }
                 pollfd.revents = pollfd.revents & !POLLIN
             }
         }
 
-        if !hangups.is_empty() {
-            self.pollfds.retain(|pollfd| !hangups.contains(&pollfd.fd));
-        }
+        // Avoid the use of `self` in closue to avoid errors due to the
+        // mutable borrow of `self` at line 483
+        let ids = &self.fdids;
```

Comment:
I think I understand the situation; I don't understand though why you consider it a problem?...

It's not exactly uncommon to create temporary bindings to get around borrow conflicts. I'm not sure this even deserves a comment: it doesn't seem like an obscure situation that requires much of an explanation?...

(BTW, including line numbers in a comment is a bad idea: these are bound to change :-) )

On the other hand, I for my part would be tempted to add a comment explaining that the entries that got removed from `fdids`, because the channels got closed, also need to be purged from `pollfds`. (And perhaps even mention that we can only do that after we finished iterating `pollfds`?...) These are the kind of things that are probably not immediately obvious even to an experienced Rust programmer looking at this code for the first time.

(I am aware that the existing `ipc-channel` code doesn't have much in the manner of such explanations... Which I personally consider a pity.)

On a more nit-picky note, I'd drop the blank line after the temporary binding -- these two lines very much belong together. Also, I think it would actually be preferable to name the temporary as `fdids` as well...